File: day19/Day19.scala
import scala.io.Source.stdin
import scala.annotation.tailrec

val PERMUTATIONS = Array(
    1, 2, 3,
    -3, 2, 1,
    -1, 2, -3,
    3, 2, -1,
    -1, -2, 3,
    -3, -2, -1,
    1, -2, -3,
    3, -2, 1,
    -2, 1, 3,
    -2, -3, 1,
    -2, -1, -3,
    -2, 3, -1,
    2, -1, 3,
    2, -3, -1,
    2, 1, -3,
    2, 3, 1,
    1, -3, 2,
    -3, -1, 2,
    -1, 3, 2,
    3, 1, 2,
    1, 3, -2,
    -3, 1, -2,
    -1, -3, -2,
    3, -1, -2,
    ).sliding(3).toList

type Vec3 = Array[Int]
type Scanner = Array[Vec3]

extension (v1: Vec3)
  def componentWise(v2: Vec3, f: (Int, Int) => Int): Vec3 =
    v1.lazyZip(v2).map{case (a, b) => f(a, b)}.toArray
  def -(v2: Vec3): Vec3 = v1.componentWise(v2, _ - _)

def find_overlap(s1: Scanner, s2: Scanner): Option[(Scanner, Vec3)] =
  {for { p <- PERMUTATIONS.iterator
        permuted = s2.map{v => p.map{i => v(i.abs - 1) * i.sign}}
        v1 <- s1.iterator
        v2 <- permuted.iterator
        diff = v2 - v1
        result = permuted.map{_ - diff}
        count = (for { a <- s1.iterator; b <- result; if a.sameElements(b) } yield 1).sum
        if count >= 12
  } yield (result, diff)
  }.nextOption

def relativize(s: IndexedSeq[Scanner]): Seq[(Scanner, Vec3)] =
  @tailrec
  def go(distances: Map[Int, (Scanner, Vec3)], next: List[Int]
                           , remaining: Set[Int]): Map[Int, (Scanner, Vec3)] =
    if next.isEmpty then
      distances
    else
      val cur :: rest = next
      val newOnes = remaining.toList.flatMap(
                      b => find_overlap(distances(cur)._1, s(b)).map{r => (b, r)})
      go(distances ++ newOnes
        , newOnes.map{_._1} ++ rest
        , remaining -- newOnes.map{_._1})

  go(Map(0 -> (s(0), Array(0, 0, 0))), List(0), Set(1 until s.length: _*)).values.toList

@main def main(idx: Int) =
  val xs = stdin.mkString.split("\n\n")
                .map{_.split('\n').drop(1)
                  .map{a => a.split(',').map{_.toInt}}}.toIndexedSeq
  val result = relativize(xs)

  println {
    if idx == 1 then
      result.flatMap{_._1.map{_.toList}}.toSet.size
    else
      val diffs = result.map{_._2}
      ( for { d1 <- diffs
              ds <- diffs.tails.to(LazyList).drop(1).dropRight(1)
              d2 <- ds
            } yield (d1 - d2).map{_.abs}.sum
      ).max
  }

File: day19/Day19.rs
#![feature(array_zip)]
use std::collections::HashMap;
use std::collections::HashSet;
use std::collections::VecDeque;
use std::env;
use std::io;

const PERMUTATIONS: [[i32; 3]; 24] = [
    [1, 2, 3],
    [-3, 2, 1],
    [-1, 2, -3],
    [3, 2, -1],
    [-1, -2, 3],
    [-3, -2, -1],
    [1, -2, -3],
    [3, -2, 1],
    [-2, 1, 3],
    [-2, -3, 1],
    [-2, -1, -3],
    [-2, 3, -1],
    [2, -1, 3],
    [2, -3, -1],
    [2, 1, -3],
    [2, 3, 1],
    [1, -3, 2],
    [-3, -1, 2],
    [-1, 3, 2],
    [3, 1, 2],
    [1, 3, -2],
    [-3, 1, -2],
    [-1, -3, -2],
    [3, -1, -2],
];

type V3 = [i32; 3];

fn main() {
    let mut scanners: Vec<Vec<V3>> =
        io::stdin()
            .lines()
            .map(Result::unwrap)
            .fold(Vec::new(), |mut res, line| {
                if line.starts_with("--") {
                    res.push(Vec::new());
                } else if !line.is_empty() {
                    let last = res.last_mut().unwrap();
                    let mut new = [0; 3];
                    line.split(',')
                        .map(|s| s.parse().unwrap())
                        .enumerate()
                        .for_each(|(i, v)| new[i] = v);
                    last.push(new);
                }

                res
            });

    let mut distances: HashMap<usize, V3> = HashMap::new();
    distances.insert(0, [0, 0, 0]);

    let mut next: VecDeque<usize> = VecDeque::new();
    next.push_back(0);

    let mut remaining: HashSet<usize> = HashSet::new();
    remaining.extend(1..scanners.len());

    while let Some(a) = next.pop_front() {
        for b in remaining.clone() {
            let ascan = &scanners[a];
            let bscan = &scanners[b];

            'outer: for p in &PERMUTATIONS {
                let idx = p.map(|i| (i.unsigned_abs() - 1) as usize);
                let sign = p.map(|i| i.signum());
                let mapper = idx.zip(sign);

                let nscan: Vec<_> = bscan
                    .iter()
                    .map(|v| mapper.map(|(i, s)| v[i] * s))
                    .collect();

                for av in ascan.iter() {
                    for nv in nscan.iter() {
                        let diff = nv.zip(*av).map(|(x, y)| x - y);

                        let temp: Vec<V3> = nscan
                            .iter()
                            .map(|v| v.zip(diff).map(|(x, y)| x - y))
                            .collect();

                        let count = ascan
                            .iter()
                            .flat_map(|v1| temp.iter().filter(move |&v2| v1 == v2))
                            .count();

                        if count >= 12 {
                            scanners[b] = temp;
                            distances.insert(b, diff);
                            next.push_back(b);
                            remaining.remove(&b);
                            break 'outer;
                        }
                    }
                }
            }
        }
    }

    if env::args().nth(1).unwrap() == "1" {
        println!(
            "{}",
            scanners.iter().flatten().collect::<HashSet<_>>().len()
        );
    } else {
        let mut max = 0;

        for i in 0..scanners.len() {
            for j in i + 1..scanners.len() {
                let manhattan = distances
                    .get(&i)
                    .and_then(|v1| {
                        distances.get(&j).map(|v2| {
                            v1.zip(*v2) //
                                .map(|(x, y)| (x - y).abs())
                                .iter()
                                .sum()
                        })
                    })
                    .unwrap_or_default();
                max = max.max(manhattan);
            }
        }

        println!("{}", max);
    }
}
